lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -139,21 +159,96 @@ def _get(
             query = query.filter(Dataset.uploaded <= json["end"])
         if "name" in json:
             query = query.filter(Dataset.name.contains(json["name"]))
+        if "filter" in json:
+            or_list = []
+            and_list = []
+            for kw in json["filter"]:
+                combine_or = False
+                contains = False
+                k, v = kw.split(":", maxsplit=1)
+                if k.startswith("^"):
+                    combine_or = True
+                    k = k[1:]
+                if v.startswith("~"):
+                    contains = True
+                    v = v[1:]
+
+                if not Metadata.is_key_path(
+                    k, Metadata.METADATA_KEYS, metalog_key_ok=True
+                ):
+                    raise APIAbort(HTTPStatus.BAD_REQUEST, str(MetadataBadKey(k)))
+                keys = k.split(".")
+                native_key = keys.pop(0).lower()
+                terms = []
+                use_dataset = False
+                dataset = None
+                user_private = None
+
+                if native_key == Metadata.DATASET:
+                    second = keys[0].lower()
+                    # The dataset namespace requires special handling because
+                    # "dataset.metalog" is really a special native key space
+                    # named "metalog", while other "dataset" sub-keys are primary
+                    # columns in the Dataset table.
+                    if second == Metadata.METALOG:
+                        native_key = keys.pop(0).lower()
+                    else:
+                        try:
+                            c = getattr(Dataset, second)
+                            column = (
+                                c
+                                if c.type.python_type is str
+                                else cast(getattr(Dataset, second), String)
+                            )
+                        except AttributeError as e:
+                            raise APIAbort(
+                                HTTPStatus.BAD_REQUEST, str(MetadataBadKey(k))
+                            ) from e
+                        use_dataset = True
+                        terms = [column.contains(v) if contains else column == v]
+                elif native_key == Metadata.USER:
+                    if not user_id:
+                        raise APIAbort(
+                            HTTPStatus.UNAUTHORIZED,
+                            f"Metadata key {k} cannot be used by an unauthenticated client",
+                        )
+                    user_private = [Metadata.user_id == user_id]
```

Comment:
I believe this is SQLAlchemy magic where this expression is not evaluated here, but is being added to a list of expressions to form the final query.  If you decide to move this "if filter" code block to a method, perhaps you could document that in its doc string?

Reply:
I wrote a fairly long and involved docstring talking about the method function, about the SQLAlchemy implementation artifacts, and about the filter syntax. Maybe it'll help.